libyul/backends/evm/ControlFlowGraph.h

```diff
@@ -0,0 +1,199 @@
+	size_t index = 0;
+	bool operator==(TemporarySlot const& _rhs) const { return &call.get() == &_rhs.call.get() && index == _rhs.index; }
+	bool operator<(TemporarySlot const& _rhs) const { return std::make_pair(&call.get(), index) < std::make_pair(&_rhs.call.get(), _rhs.index); }
+	static constexpr bool canBeFreelyGenerated = false;
+};
+/// A slot containing an arbitrary value that is always eventually popped and never used.
+/// Used to maintain stack balance on control flow joins.
+struct JunkSlot
+{
+	bool operator==(JunkSlot const&) const { return true; }
+	bool operator<(JunkSlot const&) const { return false; }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+using StackSlot = std::variant<FunctionCallReturnLabelSlot, FunctionReturnLabelSlot, VariableSlot, LiteralSlot, TemporarySlot, JunkSlot>;
+/// The stack top is usually the last element of the vector.
+using Stack = std::vector<StackSlot>;
+
+/// @returns true if @a _slot can be generated on the stack at any time.
+inline bool canBeFreelyGenerated(StackSlot const& _slot)
+{
+	return std::visit([](auto const& _typedSlot) { return std::decay_t<decltype(_typedSlot)>::canBeFreelyGenerated; }, _slot);
+}
+
+/// Control flow graph consisting of ``CFG::BasicBlock``s connected by control flow.
+struct CFG
+{
+	explicit CFG() {}
+	CFG(CFG const&) = delete;
+	CFG(CFG&&) = delete;
+	CFG& operator=(CFG const&) = delete;
+	CFG& operator=(CFG&&) = delete;
+
+	struct BuiltinCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<BuiltinFunction const> builtin;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+		/// Number of proper arguments with a position on the stack, excluding literal arguments.
+		/// Literal arguments (like the literal string in ``datasize``) do not have a location on the stack,
+		/// but are handled internally by the builtin's code generation function.
+		size_t arguments = 0;
+	};
+	struct FunctionCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<Scope::Function const> function;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+	};
+	struct Assignment
+	{
+		std::shared_ptr<DebugData const> debugData;
+		/// The variables being assigned to also occur as ``output`` in the ``Operation`` containing
+		/// the assignment, but are also stored here for convenience.
+		std::vector<VariableSlot> variables;
+	};
+
+	struct Operation
+	{
+		/// Stack slots this operation expects at the top of the stack and consumes.
+		Stack input;
+		/// Stack slots this operation leaves on the stack as output.
+		Stack output;
+		std::variant<FunctionCall, BuiltinCall, Assignment> operation;
+	};
+
+	struct FunctionInfo;
+	/// A basic control flow block containing ``Operation``s acting on the stack.
+	/// Maintains a list of entry blocks and a typed exit.
+	struct BasicBlock
+	{
+		std::vector<BasicBlock*> entries;
+		std::vector<Operation> operations;
```

Comment:
I'd keep all members grouped together. It's easy to miss some of them if they're interspersed with types. For a moment here I was wondering how it makes sense for `exit` to be the only member in this struct :)

Reply:
Fair enough :-). Changed it.